src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -871,35 +1056,43 @@ bool rasterClear(SwSurface* surface)
 
 bool rasterImage(SwSurface* surface, SwImage* image, const Matrix* transform, const SwBBox& bbox, uint32_t opacity)
 {
-    Matrix invTransform;
-
-    if (transform) {
-        if (!_inverse(transform, &invTransform)) return false;
-    }
-    else invTransform = {1, 0, 0, 0, 1, 0, 0, 0, 1};
-
     auto translucent = _translucent(surface, opacity);
 
     if (image->rle) {
         //Fast track
-        if (_identify(transform)) {
-            //OPTIMIZE ME: Support non transformed image. Only shifted image can use these routines.
-            if (translucent) return _rasterTranslucentImageRle(surface, image->rle, image->data, image->w, image->h, opacity);
-            return _rasterImageRle(surface, image->rle, image->data, image->w, image->h);
+        if (auto transformType = _identify(transform)) {
+            //Identity or NoTransformation
+            if (transformType == 1) {
+                if (translucent) return _rasterTranslucentImageRle(surface, image->rle, image->data, image->w, image->h, opacity);
+                return _rasterImageRle(surface, image->rle, image->data, image->w, image->h);
+            }
+            //Translation
+            if (translucent) return _rasterTranslucentImageRle(surface, image->rle, image->data, image->w, image->h, opacity, -transform->e13, -transform->e23);
+            return _rasterImageRle(surface, image->rle, image->data, image->w, image->h, -transform->e13, -transform->e23);
         } else {
+            Matrix invTransform;
+            if (transform && !_inverse(transform, &invTransform)) return false;
             if (translucent) return _rasterTranslucentImageRle(surface, image->rle, image->data, image->w, image->h, opacity, &invTransform);
             return _rasterImageRle(surface, image->rle, image->data, image->w, image->h, &invTransform);
         }
     }
     else {
         //Fast track
-        if (_identify(transform)) {
-            //OPTIMIZE ME: Support non transformed image. Only shifted image can use these routines.
-            if (translucent) return _rasterTranslucentImage(surface, image->data, image->w, image->h, opacity, bbox);
-            else return _rasterImage(surface, image->data, image->w, image->h, bbox);
+        if (auto transformType = _identify(transform)) {
+            //Identity or NoTransformation
+            if (transformType == 1) {
+                if (translucent) return _rasterTranslucentImage(surface, image->data, image->w, image->h, opacity, bbox);
+                else return _rasterImage(surface, image->data, image->w, image->h, bbox);
+            }
+            //Translation
+            if (translucent) return _rasterTranslucentImage(surface, image->data, image->w, image->h, opacity, bbox, -transform->e13, -transform->e23);
```

Comment:
Why don't reuse the existing method? We can just offset the region xy...
_rasterTranslucentImage(SwSurface* surface, uint32_t *img, uint32_t w, uint32_t h, uint32_t opacity, const SwBBox& region);

Reply:
@hermet I'm very sorry that it takes so long. I'll push the new patch today/tomorrow.

Reply:
my problem is that I can't see how to use a region because both pictures and shapes/svgs may use the same rastering functions. For ex an svg with clipping and opacity will use _translucentImage, and a translucent picture (pixels) uses it as well.
For a picture I can easily use region to call the same function for the identity transform and the translation, but it's not what I want to do for an svg...

Am I missing something?